keps/83-workload-preemption/README.md

```diff
@@ -0,0 +1,636 @@
+
+### Non-Goals
+
+- Graceful termination of Workloads is left to the workload pods to implement.
+- Tracking usage by workloads that take arbitrary time to be suspended. See
+  [Workload preemption doesn't imply immediate Pod termination](#workload-preemption-doesnt-imply-immediate-pod-termination) to learn more.
+  For example, the integration with Job uses the [suspend field](https://kubernetes.io/docs/concepts/workloads/controllers/job/#suspending-a-job).
+- Partial workload suspension is not supported.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+This enhancement proposes the introduction of the following APIs to influence
+Workload preemption:
+- A field in the ClusterQueue to influence how long to wait before starting
+  preemption.
+- A field in the Workload to determine whether to terminate or requeue upon preemption.
+
+And an algorithm for selecting a set of Workloads using borrowed quota to be
+preempted when a Workload in a ClusterQueue is under the min quota.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+As a cluster administrator, in the cluster I have cohorts with different
+characteristics:
+
+- In some cohorts, all the ClusterQueues are known to have short-running
+  workloads that don't do checkpointing. In this case, it's fine to wait for
+  jobs to finish, but don't allow ClusterQueues to borrow more resources.
+- In other cohorts, ClusterQueues might have long running jobs that support
+  checkpointing. In this case, it's necessary to preempt Workloads that are
+  using borrowed quota.
+
+I want to configure in the ClusterQueue how long to wait for Workloads to finish
+up to a limit.
+
+```yaml
+apiVersion: kueue.x-k8s.io/v1alpha2
+kind: ClusterQueue
+metadata:
+  name: cluster-total
+spec:
+  preemption:
+    waitBeforePreemptionSeconds: 600 # leave empty to wait indefinitely.
+```
+
+#### Story 2
+
+As a batch user, some of my workloads implement checkpointing, so I know that
+after preemption they can safely restart. Others need manual intervention before
+restart, so they should be marked as failed.
+
+This can be configured like this:
+
+```yaml
+apiVersion: kueue.x-k8s.io/v1alpha2
+kind: Workload
+metadata:
+  name: my-workload
+spec:
+  onPreemption: Requeue # OR Terminate
+```
+
+### Notes/Contraints/Caveats (Optional)
+
+#### How to configure waitBeforePreemptionSeconds
```

Comment:
I think this field still respect priority preempting on workloads, right? So, only workloads with higher priority will leverage this feature.